Show numResolved / numUnresolved in resolution view

chrisrueger opened the pull request:
- First draft to show simple counters to easier spot unresolved requirements.
- reqs resolution := optional are not counted as unresolved.

<img width="356" alt="image" src="https://github.com/bndtools/bnd/assets/188422/c15cb3f1-3168-4391-845d-ca4f98e2c706">

chrisrueger commented:
@pkriens Not sure if this is useful. Initially I thought I could use those counters it to easier find unresolved Requirements. But not sure if I am thinking correctly. 

I just added 3 counters now to count Resolved, Optional and Other. 
Originally I though "Other" == Unresolved. But not sure anymore that is the case after playing around with it. 
I thought maybe adding another filter so that you filter the Requirements list e.g. to only show the Resolved, Optional or Other. 

Any thoughts on this?

pkriens commented:
I think I'd prefer more filters ... The advantage of a graphic UI is that our brains are amazing in detecting patterns. Having more filters will allow you to use this better than numbers. And with the aggregate idea of the view, I am struggling to think what would mean?

If you're at it, it might be an idea to detect duplicate exports? It is ok if they are identical (all attributes should match) but otherwise it would help solve the problem we had.

pkriens commented:
>  find identical capabilities which are exported by > 1 resource / bundle?

No that is ok :-) the problem is when there primary attribute is the same but they differ in other attributes. Except for osgi.service namespace, the primary attribute is always the same name as the name of the namespace. E.g. for a package the namespace is "osgi.wiring.package" and the attribute "osgi.wiring.package" holds the name of the package. you can find the gory details in PackageNamespace clas. For some asinine reason we made the ServiceNamespace use "objectClass" for the primary attribute.

The problem we're chasing is finding packages that have the same name but differ in their class. Sinces the "hashes" attribute contains the hashes of each class, we'd see differences. The problem was that there were two exporters with different packages with the same name. If they'd been identical, as specified by OSGi, it would be ok.

chrisrueger commented:
> The problem we're chasing is finding packages that have the same name but differ in their class. Sinces the "hashes" attribute contains the hashes of each class, we'd see differences. The problem was that there were two exporters with different packages with the same name. If they'd been identical, as specified by OSGi, it would be ok.

Thanks @pkriens  for the clarification. 

I have pushed an update. 
Could you have a look at the [testcase](https://github.com/bndtools/bnd/pull/5923/files#diff-178fa15cdfd8915c46c01a32071df363e57b678875587de57999e8b9258520b3) to see if it makes sense?
Currently I specifically handle `osgi.wiring.package` and their `bnd.hashes`. 

**Here is how the current draft of the UI looks like:**

- added empty state short message, how to use the resolution view
- I also added two search filter inputs so that you can now search the Requirements and Capabilities similar to how you can search repos. (TODO currently I Glob-match on `req.toString()` and `cap.toString()`. That maybe too broad. let me know if you want that to be different)

<img width="1726" alt="image" src="https://github.com/bndtools/bnd/assets/188422/3c0a875a-1f73-4f37-aad5-d76f15cfe862">

- show counter of visible rows (I find this helpful to see that filters change something)
- in the Capability Panel I additionally show a "Problem"-counter (there are any). Currently a "problem" is exactly the duplicate package problem. (TODO if you have suggestions for different wording e.g. if "Problem" is too harsh...)
- there is also a new Filter button (yellow warning icon) for the "Problems"
<img width="1864" alt="image" src="https://github.com/bndtools/bnd/assets/188422/49ec2c8c-a0c9-4711-9e82-558150132299">

- pressing the new yellow warning button filters the currently visible results and only shows those who were detected as "having duplicate exports with different hashes"

<img width="1728" alt="image" src="https://github.com/bndtools/bnd/assets/188422/5fcff690-ee58-44c8-b2ff-46a9ddfb99eb">

- Double clicking on a Cap brings up the Advanced Repository Search (my previous PR)

<img width="900" alt="image" src="https://github.com/bndtools/bnd/assets/188422/2f3da6af-de4e-473c-af66-d33b6e1e273a">

<img width="320" alt="image" src="https://github.com/bndtools/bnd/assets/188422/71000da5-983e-45d3-95c6-7871b17d7e94">


Do you think this would have been helpful for debugging the "duplicate exporterts" problems?
More thoughts?

pkriens reviewed (commented):
LGTM

I think we should commit it and get some experience

pkriens commented:
Really interesting what's happening here. Maybe we should make a video to show what this is all doing.

chrisrueger commented on biz.aQute.bndlib.tests/test/aQute/bnd/osgi/resource/CapReqBuilderTest.java:
```diff
@@ -212,4 +213,50 @@ public void testCapabilityToRequirementWithFilter() throws Exception {
+
+
+
+	private CapabilityImpl createCap(String bundleSymName, String pck, String version, List<String> hashes) {
```
Done

chrisrueger replied:
FYI: Attrs.putTyped already handles arrays and collections the same. Thus I did not make the list in the method.

chrisrueger commented:
> I think we should commit it and get some experience

Thanks @pkriens for the feedback. I added the varargs suggestion in the testcase. 

Ok then let's merge :)



> Maybe we should make a video to show what this is all doing.

Sure, open to ideas. Maybe let's talk about it in the next call?
